src/core/thread/mle.cpp

```diff
@@ -739,6 +757,19 @@ otError Mle::SetMeshLocalPrefix(const uint8_t *aMeshLocalPrefix)
     netif.UnsubscribeMulticast(mLinkLocalAllThreadNodes);
     netif.UnsubscribeMulticast(mRealmLocalAllThreadNodes);
 
+#if OPENTHREAD_ENABLE_SERVICE
+
+    for (uint8_t i = 0; i < sizeof(mServiceAlocs) / sizeof(mServiceAlocs[0]); i++)
+    {
+        if (HostSwap16(mServiceAlocs[i].GetAddress().mFields.m16[7]) != Mac::kShortAddrInvalid)
+        {
```

Comment:
Maybe it would be simpler to only update the mesh local prefix part of valid Service Aloc?

Reply:
Well, I've been following the style which was there already - remove address, update structure, add address. I guess your comment could be also applied to leader aloc and mesh local address. I'm preferring it the way it is done, as it is obvious to the reader that it will work.

Reply:
The aim of (remove, update, add) is to trigger some events to notify the changes of addresses. I am not  changing this style. What I want to say is when updating Service Alocs, using `memcpy()` to update the mesh local prefix would be much more efficient than `UpdateServiceAlocs()` which would iterate through service aloc addresses and leader network data more than once. In fact, here I personally prefer updating leader aloc the same way as mesh local address. 
Thoughts?

Reply:
Right. For some reason I haven't realized earlier that I can do remove/modify/add cycle, without having to resort to `UpdateServiceAlocs`. Specifically, the list of srvs to be added is identical to the removed ones. For some reason I assumed that those lists would be different, hence `UpdateServiceAlocs` would be simplest way. But indeed, remove/update/add is much better here.

Btw. leader aloc does exactly this as well, just over few functions interleaved with `VerifyOrExit` & co.

Reply:
Yes. Leader Aloc could also be updated this way since it is in the `if (mRole == OT_DEVICE_ROLE_LEADER)` block. I am ok if you want to update it as well, or leave it as it is for now.